latency_awareness: Combat tests flakiness by not relying on timing

wprzytula opened the pull request:
The test bearing an undeniably attractive name: `latency_aware_default_policy_stops_penalising_after_min_average_increases_enough_only_after_update_rate_elapses` has shown to be flaky. I suspect it is due to the fact that it relies on timing to be valid, and execution in CI is often extremely slow. The problem, most probably, lies in minimum average updates done periodically: the test code assumes that an update has not yet been triggered at some point, which in CI happens to not be true.
To combat that:
- min avg update logic is extracted into `MinAvgUpdater` struct
- it is made configurable upon LA creation whether updates are to be scheduled to be periodically issued (as before, used in real cases) _or_ triggered manually by calls on `MinAvgUpdater`, 
- the affected tests (those that have relied on timing) are changed to use the `MinAvgUpdater` explicitly instead of scheduling periodic updates. This prevents flakiness.

Fixes: #710 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

piodul reviewed (changes requested):
I don't like this approach as it still relies on timing. You are making the test much slower to reduce the probability of failure, yet you still don't have a guarantee that a spurious failure won't occur.

I suggest changing the test case not to rely on timing at all. As I understand the problem here (btw, the PR description doesn't tell what is the _cause_ of the flakiness) is that the test waits for some period of time and hopes that a task spawned by the latency aware policy (the updater) will do some some work. Instead of waiting and praying, you can control when the updater does its work.

- Introduce a `LatencyAwareUpdater` struct with a single method `async fn tick()`,
- Introduce `fn new_for_test(...) -> (Self, LatencyAwareUpdater)` and reimplement `fn new() -> Self` using `new_for_test`,
- Use `new_for_test()` in the tests that need to control when the updater does the tick.

wprzytula commented:
> I don't like this approach as it still relies on timing. You are making the test much slower to reduce the probability of failure, yet you still don't have a guarantee that a spurious failure won't occur.
> 
> I suggest changing the test case not to rely on timing at all. As I understand the problem here (btw, the PR description doesn't tell what is the _cause_ of the flakiness) is that the test waits for some period of time and hopes that a task spawned by the latency aware policy (the updater) will do some some work. Instead of waiting and praying, you can control when the updater does its work.
> 
>     * Introduce a `LatencyAwareUpdater` struct with a single method `async fn tick()`,
> 
>     * Introduce `fn new_for_test(...) -> (Self, LatencyAwareUpdater)` and reimplement `fn new() -> Self` using `new_for_test`,
> 
>     * Use `new_for_test()` in the tests that need to control when the updater does the tick.

v2: taken the suggested approach:
- extracted update logic into `MinAvgUpdater`,
- implemented `LatencyAwareness::new_for_test()`,
- modified tests not to rely on timing, by disabling periodic min avg updates and invoking them manually instead.

piodul reviewed (approved):
LGTM. Please update the title and description of the PR because it is no longer describes what the PR really does. Then, I'll merge.
